src/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,371 @@
+#[repr(C)]
+#[derive(Copy, Clone, Debug)]
+pub struct Quaternion {
+    pub real: f64,
+    pub i: f64,
+    pub j: f64,
+    pub k: f64,
+}
+
+impl Quaternion {
+    /// Initializes a quaternion from a real component and an
+    /// imaginary 3-vector
+    pub fn new_from_vector(real: f64, imag: Vector3) -> Self {
+        Self{real: real, i: imag.x, j: imag.y, k: imag.z}
+    }
+
+    /// Initializes a Quaternion from its raw components
+    pub fn new(real: f64, i: f64, j: f64, k: f64) -> Self {
+        Self{real, i, j, k}
+    }
+
+    /// Converts from euler angles to a quaternion. The euler angles are expected to
+    /// be represented according to the Tait-Bryan formalism and applied in the Z-Y'-X"
+    /// order (where Z -> yaw, Y -> pitch, X -> roll)
+    pub fn from_euler_angles(roll: f64, pitch: f64, yaw: f64) -> Self {
+        let roll_cos = (roll * 0.5).cos();
+        let roll_sin = (roll * 0.5).sin();
+
+        let pitch_cos = (pitch * 0.5).cos();
+        let pitch_sin = (pitch * 0.5).sin();
+
+        let yaw_cos = (yaw * 0.5).cos();
+        let yaw_sin = (yaw * 0.5).sin();
+
+        let real = (roll_cos * pitch_cos * yaw_cos) + (roll_sin * pitch_sin * yaw_sin);
+        let i = (roll_sin * pitch_cos * yaw_cos) - (roll_cos * pitch_sin * yaw_sin);
+        let j = (roll_cos * pitch_sin * yaw_cos) + (roll_sin * pitch_cos * yaw_sin);
+        let k = (roll_cos * pitch_cos * yaw_sin) - (roll_sin * pitch_sin * yaw_cos);
+
+        Quaternion::new(real, i, j, k)
+    }
+
+    /// Converts a quaternion into euler angles. The euler angles are 
+    /// represented according to the Tait-Bryan formalism and applied 
+    /// in the Z-Y'-X" order (where Z -> yaw, Y -> pitch, X -> roll). 
+    /// The return value is a list of [roll, pitch, yaw]
+    pub fn to_euler_angles(&self) -> [f64;3] {
+        // get a normalized version of the quaternion
+        let quat = self.get_normalized();
+
+        // calculate yaw
+        let yaw_sin_pitch_cos = 2.0 * ((quat.real * quat.k) + (quat.i * quat.j));
+        let yaw_cos_pitch_cos = 1.0 - 2.0 * ((quat.j * quat.j) + (quat.k * quat.k));
+        let yaw = yaw_sin_pitch_cos.atan2(yaw_cos_pitch_cos);
+
+        // calculate pitch and roll
+        let pitch_sin = 2.0 * ((quat.real * quat.j) - (quat.k * quat.i));
+        let pitch: f64;
+        let roll: f64;
+        // for a pitch that is π / 2, we experience gimbal lock
+        // and must calculate roll based on the real rotation and yaw
+        if pitch_sin.abs() >= 1.0 {
+            pitch = (std::f64::consts::PI / 2.0).copysign(pitch_sin);
+            roll = (2.0 * quat.i.atan2(quat.real)) + yaw.copysign(pitch_sin);
+        } else {
+            pitch = pitch_sin.asin();
+            let roll_sin_pitch_cos = 2.0 * ((quat.real * quat.i) + (quat.j * quat.k));
+            let roll_cos_pitch_cos = 1.0 - 2.0 * ((quat.i * quat.i) + (quat.j * quat.j));
+            roll = roll_sin_pitch_cos.atan2(roll_cos_pitch_cos);
+        }
+
+
```

Comment:
Mild preference to get rid of one of these blank lines. One blank line is enough inside a function.